File: Data Structures & Algorithms/binary-tree-diameter/submission-5.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:
    def diameterOfBinaryTree(self, root: Optional[TreeNode]) -> int:
        res = 0

        # Returns height
        def dfs(node):
            if not node:
                return 0

            print(f"entered node: {node.val}")

            nonlocal res
            left, right = dfs(node.left), dfs(node.right)
            print(f"current node: {node.val}, left: {left}, right: {right}")

            # set res to max of itself and current diameter (left + right)
            res = max(res, left + right)
            print(f"res: {res}")

            # return height (max of left and right subtrees) + 1 for current node
            return 1 + max(left, right)

        dfs(root)

        return res

File: Data Structures & Algorithms/search-for-word/submission-5.cpp
class Solution {
    int m, n;
    vector<vector<char>> board;
    string word;

    bool dfs(int r, int c, int i) {
        if (i == word.size()) {
            return true;
        }
        if (r < 0 || c < 0 || r == m || c == n || board[r][c] != word[i] || board[r][c] == '#') {
            return false;
        }

        board[r][c] = '#';
        bool res = dfs(r + 1, c, i + 1) || dfs(r - 1, c, i + 1) || dfs(r, c + 1, i + 1) ||
                   dfs(r, c - 1, i + 1);
        board[r][c] = word[i];
        return res;
    }

   public:
    bool exist(vector<vector<char>>& board, string word) {
        m = board.size(), n = board[0].size();
        this->board = board;
        this->word = word;

        for (int r = 0; r < m; ++r) {
            for (int c = 0; c < n; ++c) {
                if (dfs(r, c, 0)) {
                    return true;
                }
            }
        }
        return false;
    }
};
